lib/Dialect/FIRRTL/Analysis/DetectCombLoops.cpp

```diff
@@ -0,0 +1,228 @@
+  for (auto connect : module.getOps<ConnectOp>())
+    DFSVisitValue(connect.dest());
+}
+
+/// DFS visit a value. Return the minimum DFS index in the SCC.
+int64_t Detector::DFSVisitValue(Value value) {
+  if (valueIndexMap.count(value))
+    return valueIndexMap[value];
+
+  // Record the minimum index in the SCC.
+  auto minIndex = index;
+
+  // Get the index of the current value and increase the global index.
+  valueIndexMap[value] = index++;
+
+  // Early return when meeting a register.
+  auto definingOp = value.getDefiningOp();
+  if (definingOp && isa<RegOp, RegResetOp>(definingOp))
+    return minIndex;
+
+  // If the current value is a module argument and not equal to `currentArg`, we
+  // should record this path as a combinational path.
+  if (!definingOp && currentArg && currentArg != value) {
+    auto outputIndex = BlockArgument(value.getImpl()).getArgNumber();
+    combPaths[currentArg.getArgNumber()].push_back(outputIndex);
+  }
+
+  auto currentStackSize = valueStack.size();
+  valueStack.push_back(value);
+
+  // Collect childs of the current value.
+  SmallVector<Value> childs;
+  for (auto user : value.getUsers()) {
+    llvm::append_range(childs, user->getResults());
+    // If the current value is the `src` of a connect operation, the `dest`
+    // value should be considered as a child.
+    if (auto connect = dyn_cast<ConnectOp>(user)) {
+      if (value != connect.dest())
+        childs.push_back(connect.dest());
+    }
+  }
+
+  // Handle two special operations that may introduce extra childs.
+  if (definingOp) {
+    // Handle instance operation.
+    if (auto instance = dyn_cast<InstanceOp>(definingOp)) {
+      if (auto referencedModule =
+              dyn_cast<FModuleOp>(instance.getReferencedModule())) {
+        // TODO: We should have a InstantiationAnalysis to construct an
+        // instantiation graph for FIRRTL or HW dialect. As a result, we can
+        // traverse the modules in a DFS order and avoid the logic here.
+        if (!combPathsMap.count(referencedModule))
+          Detector(referencedModule, combPathsMap).detect();
+
+        // Append all instance outputs that are combinationally connected to the
+        // current value.
+        auto valueIndex = OpResult(value.getImpl()).getResultNumber();
+        for (auto outputIndex : combPathsMap[referencedModule][valueIndex])
+          childs.push_back(instance.getResult(outputIndex));
+      }
+    }
+
+    // Handle subfield operation. After lower types, subfield operations should
+    // only be used to split bundles generated by memories. Thus we need to
+    // detect whether there are combinational path between inputs (`en` and
+    // `addr`) and outputs (`data` or `rdata`).
+    if (auto subField = dyn_cast<SubfieldOp>(definingOp)) {
+      auto inputDefiningOp = subField.input().getDefiningOp();
+      if (!inputDefiningOp)
+        return subField.emitOpError("input must not be an argument"), minIndex;
```

Comment:
Why not?  What phase of FIRRTL lowering does this work on?

Please make the error message more explicit about what the problem is, particularly if this is user-visible.